### ISSUE TITLE
Index not stored when closing file when duckdb version < 0.4.1

### ISSUE THREAD

**Issue by andreas-wolf** [duckdb/duckdb#4126]
#### What happens?
TLDR: This bug seems to be fixed with the latest master version duckdb-0.4.1.dev710
Since I spent more than a day to find the reason why my code was not working as expected and I did not find this behavior documented anywhere I'd like to report it anyway for documentary reasons. Feel free to close it immediately, but please leave it for others as reference since right now the current version 0.4.0 on PyPI still shows this behavior.

After closing the database the database index is gone.
I would expect that the index is still there after reopening the database.
I stumbled over this behaviour because UNIQUE INDEX checks were missing after reopening the db.

#### To Reproduce
Here is some python sample code. I can reproduce the bug with version duckdb-0.3.2 and duckdb-0.4.0 but not with duckdb-0.4.1.dev710
If you run the code there is a schema.sql in the subfolder schema_ok containing the index and the same database exported to schema_wrong after reopening the connection. The index is missing now.
```
# %%
import os
import duckdb

# %%
def create_table(
    con,
):
    sql = f"""
DROP TABLE IF EXISTS raw;
DROP INDEX IF EXISTS customer_year_month_idx;
CREATE TABLE raw(
    "year" SMALLINT,
    "month" TINYINT,
    "day" TINYINT,
    "customer_ID" BIGINT
);
CREATE UNIQUE INDEX customer_year_month_idx ON raw (customer_ID, year, month);
"""
    con.execute(sql)


# %%
def create_dir(dir_name):
    if not os.path.exists(dir_name):
        os.makedirs(dir_name)


# %%
def main():
    for dir in ["schema_ok", "schema_wrong"]:
        create_dir(dir)
    database = "test.duckdb"
    con = duckdb.connect(database=database)
    create_table(con)
    con.execute("EXPORT DATABASE 'schema_ok';")
    con.close()
    # Now the index is missing from the database.
    con = duckdb.connect(database=database)
    con.execute("EXPORT DATABASE 'schema_wrong';")
    con.close()


# %%
if __name__ == "__main__":
    main()
```
#### Environment (please complete the following information):
 - OS: Ubuntu 22.04 LTS
 - DuckDB Version: duckdb-0.4.0
 - DuckDB Client: Python 3.9.13

#### Identity Disclosure:
 - Full Name: Andreas Wolf
 - Affiliation: ChromaNet

If the above is not given and is not obvious from your GitHub profile page, we might close your issue without further review. Please refer to the [reasoning behind this rule](https://berthub.eu/articles/posts/anonymous-help/) if you have questions.

#### Before Submitting

- [ x] **Have you tried this on the latest `master` branch?**
* **Python**: `pip install duckdb --upgrade --pre`
* **R**: `install.packages("https://github.com/duckdb/duckdb/releases/download/master-builds/duckdb_r_src.tar.gz", repos = NULL)`
* **Other Platforms**: You can find binaries [here](https://github.com/duckdb/duckdb/releases/tag/master-builds) or compile from source.

- [x ] **Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?**

**Comment by Mytherin** [comment]
Thanks for the report!

Indexes were not persisted in prior versions of DuckDB. This behavior was expected and [also documented](https://duckdb.org/docs/sql/indexes). This has been fixed in the latest master by @pdet in #4022, and persistent indexes will be part of the next release.

**Comment by andreas-wolf** [comment]
Ah, maybe I boiled down the problem a bit too much or I don't understand the meaning of that sentence in doc 0.4.0

> Currently, ART indexes are not persisted. Unique and primary key indexes are rebuilt upon startup, while user-defined indexes are discarded.

I understood that when the database is opened again after closing the unique index should be rebuild, which was not the case.
When I closed the connection and reopened it, there was no unique index anymore.
And sureley the unique index should have been mentioned in the schema export I thought, because how could it be rebuild if it's no longer in the schema?

Nevertheless, it seems fixed now.
Thanks for your remarkable fast reaction.

**Comment by Mytherin** [comment]
Ah yes, you are correct that there might have been a bug where explicitly created unique indices (as opposed to implicitly created through defining a UNIQUE constraint) were not preserved. But indeed this seems to be resolved now.

**Comment by korpa** [comment]
@Mytherin  I think it is not resolved, yet. I use v0.6.1 CLI. 

If I create an unique index after creating a table via a shell script like so

```shell
cat << EOF | duckdb ./orders.db
CREATE UNIQUE INDEX orders_ordernumber_idx_unique ON orders (ordernumber);`
EOF
```

the index is not preserved

**Comment by Mytherin** [comment]
That is probably related to #5984. Does it work if you run `CHECKPOINT` after the `CREATE INDEX` statement?
